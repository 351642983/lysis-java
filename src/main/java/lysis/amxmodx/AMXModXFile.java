package lysis.amxmodx;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.zip.Inflater;

import lysis.BitConverter;
import lysis.ExtendedDataInputStream;
import lysis.PawnFile;
import lysis.Public;
import lysis.lstructure.Argument;
import lysis.lstructure.Dimension;
import lysis.lstructure.Function;
import lysis.lstructure.Native;
import lysis.lstructure.Scope;
import lysis.lstructure.Tag;
import lysis.lstructure.Variable;
import lysis.lstructure.VariableType;

public class AMXModXFile extends PawnFile {

	public final static long MAGIC2 = 0x414D5858;
	public final static long MAGIC2_VERSION = 0x0300;
	public final static long AMX_MAGIC = 0xE0F1;
	public final static long AMX_DBG_MAGIC = 0xEFF1;
	public final static long MIN_FILE_VERSION = 6;
	public final static long MIN_DEBUG_FILE_VERSION = 8;
	public final static long CUR_FILE_VERSION = 8;
	public final static int DEFSIZE = 8;
	public final static int AMX_FLAG_DEBUG = 0x2;
	public final static byte IDENT_VARIABLE = 1;
	public final static byte IDENT_REFERENCE = 2;
	public final static byte IDENT_ARRAY = 3;
	public final static byte IDENT_REFARRAY = 4;
	public final static byte IDENT_FUNCTION = 9;
	public final static byte IDENT_VARARGS = 11;

	private final static long Q_USER_TAG_STRING = 100500;

	public class PluginHeader {
		public byte cellsize;
		public int disksize;
		public int imagesize;
		public int memsize;
		public int offset;
	}

	private class AMX_HEADER {
		public int size;
		public int magic;
		public byte file_version;
		public byte amx_version;
		public short flags;
		public short defsize;
		public int cod;
		public int dat;
		public int hea;
		public int stp;
		public int cip;
		public int publics;
		public int natives;
		public int libraries;
		public int pubvars;
		public int tags;
		public int nametable;
	}

	private class AMX_DEBUG_HDR {
		public int size;
		public int magic;
		public byte file_version;
		public byte amx_version;
		public short flags;
		public short files;
		public short lines;
		public short symbols;
		public short tags;
		public short automatons;
		public short states;

		public final static int SIZE = 4 + 2 + (1 * 2) + (4 * 7);
	}

	private byte[] DAT_;
	private Variable[] allvars_;
	private Tag stringTag;
	private Automation[] automations_;
	private State[] states_;

	public AMXModXFile(byte[] binary) throws Exception {
		ExtendedDataInputStream reader = new ExtendedDataInputStream(new ByteArrayInputStream(binary));
		long magic = reader.ReadUInt32();

		if (magic == MAGIC2) {
			int version = reader.ReadUInt16();
			if (version > MAGIC2_VERSION)
				throw new Exception("unexpected version");

			PluginHeader ph = null;
			byte numPlugins = reader.readByte();
			for (byte i = 0; i < numPlugins; i++) {
				PluginHeader p = new PluginHeader();
				p.cellsize = reader.readByte();
				p.disksize = reader.ReadInt32();
				p.imagesize = reader.ReadInt32();
				p.memsize = reader.ReadInt32();
				p.offset = reader.ReadInt32();
				if (p.cellsize == 4) {
					ph = p;
					break;
				}
			}
			if (ph == null)
				throw new Exception("could not find applicable cell size");

			int bufferSize = ph.imagesize > ph.memsize ? ph.imagesize + 1 : ph.memsize + 1;
			byte[] bits = new byte[bufferSize];

			Inflater gzip = new Inflater(true);
			gzip.setInput(binary, ph.offset + 2, ph.disksize - 2);
			int read = gzip.inflate(bits, 0, bufferSize);
			if (read != ph.imagesize)
				System.err.printf("uncompressed size mismatch, bad file? expected %d, got %d\n", ph.imagesize, read);

			binary = bits;
		} else {
			throw new Exception("unrecognized file");
		}

		reader.close();
		reader = new ExtendedDataInputStream(new ByteArrayInputStream(binary));
		AMX_HEADER amx = new AMX_HEADER();
		amx.size = reader.ReadInt32();
		amx.magic = reader.ReadUInt16();
		amx.file_version = reader.readByte();
		amx.amx_version = reader.readByte();
		amx.flags = reader.ReadInt16();
		amx.defsize = reader.ReadInt16();
		amx.cod = reader.ReadInt32();
		amx.dat = reader.ReadInt32();
		amx.hea = reader.ReadInt32();
		amx.stp = reader.ReadInt32();
		amx.cip = reader.ReadInt32();
		amx.publics = reader.ReadInt32();
		amx.natives = reader.ReadInt32();
		amx.libraries = reader.ReadInt32();
		amx.pubvars = reader.ReadInt32();
		amx.tags = reader.ReadInt32();
		amx.nametable = reader.ReadInt32();

		if (amx.magic != AMX_MAGIC)
			throw new Exception(String.format("unrecognized amx header %x", amx.magic));

		if (amx.file_version < MIN_FILE_VERSION || amx.file_version > CUR_FILE_VERSION) {
			throw new Exception("unrecognized amx version");
		}

		if (amx.defsize != DEFSIZE)
			throw new Exception("unrecognized header defsize");

		DAT_ = new byte[amx.hea - amx.dat];
		for (int i = 0; i < DAT_.length; i++)
			DAT_[i] = binary[amx.dat + i];

		if (amx.publics > 0) {
			int count = (amx.natives - amx.publics) / DEFSIZE;
			publics_ = new Public[count];
			ExtendedDataInputStream r = new ExtendedDataInputStream(
					new ByteArrayInputStream(binary, amx.publics, count * DEFSIZE));
			for (int i = 0; i < publics_.length; i++) {
				long address = r.ReadUInt32();
				int nameoffset = r.ReadInt32();
				String name = ReadName(binary, nameoffset);
				publics_[i] = new Public(name, address);
			}
			r.close();
		}

		// .CODE
		{
			byte[] codeBytes = Slice(binary, amx.cod, amx.dat - amx.cod);
			code_ = new Code(codeBytes, (byte) 0, 0, 0, 0);
		}

		// .DATA
		{
			byte[] dataBytes = Slice(binary, amx.dat, amx.size - amx.dat);
			data_ = new Data(dataBytes, amx.size - amx.dat);
		}

		// .NATIVES
		if (amx.natives > 0) {
			int count = (amx.libraries - amx.natives) / DEFSIZE;
			natives_ = new Native[count];
			ExtendedDataInputStream r = new ExtendedDataInputStream(
					new ByteArrayInputStream(binary, amx.natives, count * DEFSIZE));
			for (int i = 0; i < count; i++) {
				r.ReadUInt32(); // Address
				int nameoffset = r.ReadInt32();
				String name = ReadName(binary, nameoffset);
				natives_[i] = new Native(name, i);
			}
			r.close();
		}

		// .PUBVARS
		if (amx.pubvars > 0) {
			int count = (amx.tags - amx.pubvars) / DEFSIZE;
			pubvars_ = new PubVar[count];
			ExtendedDataInputStream r = new ExtendedDataInputStream(
					new ByteArrayInputStream(binary, amx.pubvars, count * DEFSIZE));
			for (int i = 0; i < count; i++) {
				long address = r.ReadUInt32();
				int nameoffset = r.ReadInt32();
				String name = ReadName(binary, nameoffset);
				pubvars_[i] = new PubVar(name, address);
			}
			r.close();
		}

		// Debug stuff.
		if (amx.file_version >= MIN_DEBUG_FILE_VERSION && (amx.flags & AMX_FLAG_DEBUG) == AMX_FLAG_DEBUG) {
			int debugOffset = amx.size;
			ExtendedDataInputStream r = new ExtendedDataInputStream(
					new ByteArrayInputStream(binary, debugOffset, AMX_DEBUG_HDR.SIZE));
			AMX_DEBUG_HDR dbg = new AMX_DEBUG_HDR();
			dbg.size = r.ReadInt32();
			dbg.magic = r.ReadUInt16();
			dbg.file_version = r.readByte();
			dbg.amx_version = r.readByte();
			dbg.flags = r.ReadInt16();
			dbg.files = r.ReadInt16();
			dbg.lines = r.ReadInt16();
			dbg.symbols = r.ReadInt16();
			dbg.tags = r.ReadInt16();
			dbg.automatons = r.ReadInt16();
			dbg.states = r.ReadInt16();

			if (dbg.magic != AMX_DBG_MAGIC)
				throw new Exception("unrecognized debug magic");

			r = new ExtendedDataInputStream(
					new ByteArrayInputStream(binary, debugOffset + AMX_DEBUG_HDR.SIZE, dbg.size - AMX_DEBUG_HDR.SIZE));

			// Fill sp struct
			debugHeader_.numFiles = dbg.files;
			debugHeader_.numLines = dbg.lines;
			debugHeader_.numSyms = dbg.symbols;

			// Read files.
			debugFiles_ = new DebugFile[dbg.files];
			for (short i = 0; i < dbg.files; i++) {
				long offset = r.ReadUInt32();
				String name = ReadName(r);
				debugFiles_[i] = new DebugFile(name, offset);
				// System.out.println(i + ": " + debugFiles_[i]);
			}

			// Read lines.
			debugLines_ = new DebugLine[dbg.lines];
			for (short i = 0; i < dbg.lines; i++) {
				long offset = r.ReadUInt32();
				int lineno = r.ReadInt32();
				debugLines_[i] = new DebugLine(lineno, offset);
				// System.out.println(i + ": " + debugLines_[i]);
			}

			List<Function> functions = new LinkedList<Function>();
			List<Variable> globals = new LinkedList<Variable>();
			List<Variable> locals = new LinkedList<Variable>();
			List<Variable> allvars = new LinkedList<Variable>();

			// Read symbols.
			for (short i = 0; i < dbg.symbols; i++) {
				int addr = r.ReadInt32();
				int tagid = r.ReadUInt16();
				long codestart = r.ReadUInt32();
				long codeend = r.ReadUInt32();
				byte ident = r.readByte();
				byte vclassByte = r.readByte();
				Scope vclass = Scope.Local;
				if (vclassByte >= 0 && vclassByte < Scope.values().length)
					vclass = Scope.values()[vclassByte];
				int dimcount = r.ReadInt16();
				String name = ReadName(r);
				// System.out.printf("addr %x, tagid %x, codestart %x, codeend %x, ident %x,
				// vclass %x, dimcount %d, name %s%n", addr, tagid, codestart, codeend, ident,
				// vclassByte, dimcount, name);

				if (ident == IDENT_FUNCTION) {
					Function func = new Function(addr, codestart, codeend, name, tagid);
					functions.add(func);
				} else {
					VariableType type = FromIdent(ident);
					Dimension[] dims = null;
					if (dimcount > 0) {
						dims = new Dimension[dimcount];
						for (int j = 0; j < dimcount; j++) {
							short tag_id = r.ReadInt16();
							int size = r.ReadInt32();
							dims[j] = new Dimension(tag_id, null, size);
						}
					}

					Variable var = new Variable(addr, tagid, null, codestart, codeend, type, vclass, name, dims);
					if (vclass == Scope.Global)
						globals.add(var);
					else
						locals.add(var);
				}
			}

			Collections.sort(globals, new Comparator<Variable>() {
				@Override
				public int compare(Variable var1, Variable var2) {
					return (int) (var1.address() - var2.address());
				}
			});
			Collections.sort(functions, new Comparator<Function>() {
				@Override
				public int compare(Function fun1, Function fun2) {
					return (int) (fun1.address() - fun2.address());
				}
			});

			allvars.addAll(locals);
			allvars.addAll(globals);
			Collections.sort(allvars, new Comparator<Variable>() {
				@Override
				public int compare(Variable var1, Variable var2) {
					return (int) (var1.address() - var2.address());
				}
			});

			variables_ = locals.toArray(new Variable[0]);
			globals_ = globals.toArray(new Variable[0]);
			functions_ = functions.toArray(new Function[0]);
			allvars_ = allvars.toArray(new Variable[0]);

			// Find tags.
			tags_ = new Tag[dbg.tags + 1];
			for (short i = 0; i < dbg.tags; i++) {
				int tag_id = r.ReadUInt16();
				String name = ReadName(r);
				tags_[i] = new Tag(name, tag_id);

				// System.out.printf("%d: %s%n", i, tags_[i]);
			}

			// Automations
			automations_ = new Automation[dbg.automatons];
			for (short i = 0; i < dbg.automatons; i++) {
				short automation_id = r.ReadInt16();
				int addr = r.ReadInt32(); // address of state variable
				String name = ReadName(r);
				automations_[i] = new Automation(automation_id, addr, name);

				// System.out.printf("%d: %s%n", i, automations_[i]);
			}

			// States.
			states_ = new State[dbg.states];
			for (short i = 0; i < dbg.states; i++) {
				short state_id = r.ReadInt16();
				short automation_id = r.ReadInt16();
				String name = ReadName(r);
				states_[i] = new State(state_id, automation_id, name);

				// System.out.printf("%d: %s%n", i, states_[i]);
			}

			stringTag = new Tag("String", Q_USER_TAG_STRING);
			tags_[dbg.tags] = stringTag;

			// Update symbols.
			for (int i = 0; i < functions_.length; i++)
				functions_[i].setTag(findTag(functions_[i].tag_id()));
			for (int i = 0; i < variables_.length; i++)
				variables_[i].setTag(findTag(variables_[i].tag_id(), variables_[i]));
			for (int i = 0; i < globals_.length; i++)
				globals_[i].setTag(findTag(globals_[i].tag_id(), globals_[i]));

			// For every function, attempt to build argument information.
			for (int i = 0; i < functions_.length; i++) {
				Function fun = functions_[i];
				int argOffset = 12;
				LinkedList<Argument> args = new LinkedList<Argument>();
				do {
					Variable var = lookupVariable(fun.address(), argOffset);
					if (var == null)
						break;

					// Add string tag to possible string arguments.
					if (var.type() == VariableType.ArrayReference && var.dims() != null && var.dims().length == 1
							&& var.dims()[0].size() == 0 && var.tag() != null && var.tag().name() == "_") {
						var.setTag(stringTag);
						var.setTagId(Q_USER_TAG_STRING);
					}

					Argument arg = new Argument(var.type(), var.name(), (int) var.tag().tag_id(), var.tag(),
							var.dims());
					args.add(arg);
					argOffset += 4;
				} while (true);
				fun.setArguments(args);
			}
		}
	}

	private static String ReadName(byte[] bytes, int offset) {
		int count = offset;
		for (; count < bytes.length; count++) {
			if (bytes[count] == 0)
				break;
		}
		try {
			return new String(bytes, offset, count - offset, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}

		return null;
	}

	private static String ReadName(DataInputStream r) throws IOException {
		LinkedList<Byte> buffer = new LinkedList<>();
		do {
			byte b = r.readByte();
			if (b == 0)
				break;
			buffer.add(b);
		} while (true);

		try {
			byte[] bytes = new byte[buffer.size()];
			for (int i = 0; i < bytes.length; i++)
				bytes[i] = buffer.get(i);

			return new String(bytes, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}

		return null;
	}

	private static String ReadString(byte[] bytes, int offset) {
		List<Byte> buffer = new LinkedList<>();
		int count = offset;
		for (; count < bytes.length; count += 4) {
			if (bytes[count] == 0)
				break;
			int cell = BitConverter.ToInt32(bytes, count);
			buffer.add((byte) cell);
		}
		try {
			byte[] chars = new byte[buffer.size()];
			for (int i = 0; i < chars.length; i++)
				chars[i] = buffer.get(i);

			return new String(chars, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}

		return null;
	}

	private static String ReadStringEx(byte[] bytes, int offset, int maxread) {
		List<Byte> buffer = new LinkedList<>();
		int count = offset;
		int last = count + maxread;
		for (; count < bytes.length && count <= last; count += 4) {
			if (bytes[count] == 0)
				break;
			int cell = BitConverter.ToInt32(bytes, count);
			buffer.add((byte) cell);
		}

		try {
			byte[] chars = new byte[buffer.size()];
			for (int i = 0; i < chars.length; i++)
				chars[i] = buffer.get(i);

			return new String(chars, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}

		return null;
	}

	private static VariableType FromIdent(byte ident) {
		switch (ident) {
		case IDENT_VARIABLE:
			return VariableType.Normal;
		case IDENT_REFERENCE:
			return VariableType.Reference;
		case IDENT_ARRAY:
			return VariableType.Array;
		case IDENT_REFARRAY:
			return VariableType.ArrayReference;
		case IDENT_VARARGS:
			return VariableType.Variadic;
		default:
			return VariableType.Normal;
		}
	}

	private Tag findTag(long tag_id, Variable var) {
		Tag tag = findTag(tag_id);

		if ("_".equals(tag.name()))
			return tag;

		Tag maybe = findTagString(var);
		if (maybe != null)
			return maybe;

		return tag;
	}

	private Tag findTagString(Variable var) {
		if (var.scope() == Scope.Local)
			return null;

		if (var.dims() == null)
			return null;

		if (var.dims().length == 1 && (var.type() == VariableType.ArrayReference || var.type() == VariableType.Array
				|| var.type() == VariableType.Reference)) {

			long size = 0;
			for (int i = 0; i < allvars_.length - 1; i++) {
				if (allvars_[i] != var)
					continue;

				size = allvars_[i + 1].address() - var.address();
			}

			// last string not detected
			if (size <= 0)
				return null;

			int end = (int) (var.address() + size - 1);

			if (DAT_[end] != 0)
				return null;

			// See if it contains only valid characters.
			int addr = (int) var.address();
			for (; addr < DAT_.length && addr < end && DAT_[addr] != 0; addr += 4) {
				int cell = BitConverter.ToInt32(DAT_, addr);
				if (!Character.isValidCodePoint(cell))
					return null;
			}

			// Null character in the middle of the string.
			if (addr != end)
				return null;

			return stringTag;
		}

		return null;
	}

	@Override
	public String stringFromData(long address) {
		return ReadString(DAT_, (int) address);
	}

	@Override
	public String stringFromData(long address, int maxread) {
		return ReadStringEx(DAT_, (int) address, maxread);
	}

	@Override
	public float floatFromData(long address) {
		return BitConverter.ToSingle(DAT_, address);
	}

	@Override
	public int int32FromData(long address) {
		assert (address >= 0 && DAT_.length > address);
		if (address >= 0 && DAT_.length <= address)
			return 0;
		return BitConverter.ToInt32(DAT_, (int) address);
	}

	@Override
	public byte[] DAT() {
		return DAT_;
	}

	@Override
	public boolean PassArgCountAsSize() {
		return true;
	}

	@Override
	public Automation lookupAutomation(long state_addr) {
		for (Automation auto : automations_) {
			if (auto.address() == state_addr)
				return auto;
		}
		return null;
	}

	@Override
	public String lookupState(short state_id, short automation_id) {
		for (State state : states_) {
			if (state.automation_id() == automation_id && state.state_id() == state_id)
				return state.name();
		}
		return null;
	}

}
