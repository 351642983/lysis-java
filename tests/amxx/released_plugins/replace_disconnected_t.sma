#include <amxmodx>
#include <amxmisc>
#include <cstrike>
#include <fakemeta>
#include <hamsandwich>

#define TID_RESTORE_ORIGIN 457547

new Float:_pg_ftmp
#define _Vec3ToAngles(%1,%2) _pg_ftmp = floatsqroot(%1[0]*%1[0] + %1[1]*%1[1]);\
    %2[1] = floatacos(%1[0]/_pg_ftmp, 1)*(1-2*_:(%1[1]<0));\
    %2[0] =-floatatan(%1[2]/_pg_ftmp, 1);\
    %2[2] = 0.0
    
#define _Vec3AddScalar(%1,%2) %1[0]+=%2;%1[1]+=%2;%1[2]+=%2
#define _Vec3MultScalar(%1,%2) %1[0]*=%2;%1[1]*=%2;%1[2]*=%2
#define _Vec3Set(%1,%2) %1[0]=%2[0];%1[1]=%2[1];%1[2]=%2[2]
#define _Vec3Add(%1,%2) %1[0]+=%2[0];%1[1]+=%2[1];%1[2]+=%2[2]    

#define _Set(%1,%2) %1|=1<<%2
#define _UnSet(%1,%2) %1&=~(1<<%2)
#define _Is(%1,%2) (%1&1<<%2)
new _human
new _alive

#define _IsPlayer(%1) (1<=%1<=g_max_players)


new g_max_players
new g_msg_saytext
new Float:g_t_origin[3]
new Float:g_b_origin[3]
new bool:g_valid_origin
new bool:g_ml


public plugin_init(){
    register_plugin("Replace Disconnected T", "1.5", "Sylwester")
    RegisterHam(Ham_Use, "func_button", "button_pushed", 1)
    RegisterHam(Ham_Spawn, "player", "Player_Spawn", 1)
    RegisterHam(Ham_Killed, "player", "Player_Killed", 1)
    g_max_players = get_maxplayers()
    g_msg_saytext = get_user_msgid("SayText")
    
    new path[128]
    get_datadir(path, 127)
    add(path, 127, "/lang/replace_disconnected_t.txt")
    if(file_exists(path)){
        register_dictionary("replace_disconnected_t.txt")
        g_ml = true
    }
}


public button_pushed(ent, idcaller, idactivator, use_type, Float:value){
    if(!_IsPlayer(idcaller) || !_Is(_alive, idcaller) || !pev_valid(ent) || cs_get_user_team(idcaller) != CS_TEAM_T)
        return
    new Float:ftmp[3]
    pev(ent, pev_mins, g_b_origin)
    pev(ent, pev_maxs, ftmp)

    _Vec3Add(g_b_origin, ftmp)
    _Vec3MultScalar(g_b_origin, 0.5) 

    pev(idcaller, pev_origin, g_t_origin)
    if(pev(idcaller, pev_flags) & FL_DUCKING)
        g_t_origin[2] += 19
    g_valid_origin = true
}


public Player_Killed(id){
    _UnSet(_alive, id)
}


public Player_Spawn(id){
    if(!is_user_alive(id))
        return
    _Set(_alive, id)
    if(cs_get_user_team(id) != CS_TEAM_T)
        return
    g_valid_origin = false
}


public client_putinserver(id){
    if(is_user_bot(id) || is_user_hltv(id)) 
        return
    _Set(_human, id)
}


public announce_t_change(newid, oldid){
    new msg[160], otname[32], ntname[32]
    get_user_name(oldid, otname, 31)
    get_user_name(newid, ntname, 31)
    if(!g_ml){
        formatex(msg, 159, "^3[ DeathRun ] ^4%s^1 has left the game. ^4%s^1 is now the terrorist.", otname, ntname)
        message_begin(MSG_ALL, g_msg_saytext, _, 0)
        write_byte(oldid)
        write_string(msg)
        message_end()    
        return
    }
    for(new i=1; i<=g_max_players; i++){
        if(!_Is(_human, i))
            continue
        formatex(msg, 159, "%L", i, "ANN_REPL_T", otname, ntname)         
        message_begin(MSG_ONE, g_msg_saytext, _, i)
        write_byte(oldid)
        write_string(msg)
        message_end()
    }
}


public client_disconnect(id){  
    _UnSet(_human, id)
    _UnSet(_alive, id)
    new players[32], pnum
    for(new i=1; i<=g_max_players; i++){
        if(!_Is(_human, i))
            continue
        if(cs_get_user_team(i) == CS_TEAM_T)
            return
        players[pnum++] = i
    }
    if(pnum<=0)
        return
    new new_terr = players[random(pnum)]
    cs_set_user_team(new_terr, CS_TEAM_T)

    announce_t_change(new_terr, id)

    if(!g_valid_origin){
        ExecuteHamB(Ham_CS_RoundRespawn, new_terr)
        return
    }
    ExecuteHamB(Ham_CS_RoundRespawn, new_terr)
    g_valid_origin = true
    new param[2]
    param[0] = new_terr
    param[1] = 10 //max teleport attempts
    restore_origin(param)
}


public restore_origin(param[]) {
    if(!g_valid_origin || --param[1]<0 || !_Is(_alive, param[0]))
        return
    new tr, Float:ftmp[3]
    _Vec3Set(ftmp, g_t_origin)
    
    //find vacant location
    for(new i=0; i<100; i++){      
        engfunc(EngFunc_TraceHull, ftmp, ftmp, 0, HULL_HUMAN, 0, tr)
        if(get_tr2(tr, TR_StartSolid) || get_tr2(tr, TR_AllSolid)){
            _Vec3AddScalar(ftmp, random_float(-10.0, 10.0))
            continue
        }   
        //check if new location is on the correct side of the wall
        engfunc(EngFunc_TraceLine, g_t_origin, ftmp, DONT_IGNORE_MONSTERS, 0, tr)
        new Float:fraction
        get_tr2(tr, TR_flFraction, fraction)
        if(fraction != 1.0){
            break
        }
        _Vec3Set(g_t_origin, ftmp)     
        
        set_pev(param[0], pev_origin, g_t_origin) //move player to vacant location
       
        //calculate angles
        _Vec3Set(ftmp, -g_t_origin)        
        _Vec3Add(ftmp, g_b_origin)
        _Vec3ToAngles(ftmp, ftmp)
        
        //aim at button
        set_pev(param[0], pev_angles, ftmp)
        set_pev(param[0], pev_fixangle, 1)
        return
    }
    set_task(0.1, "restore_origin", TID_RESTORE_ORIGIN+param[0], param, 2)
}
