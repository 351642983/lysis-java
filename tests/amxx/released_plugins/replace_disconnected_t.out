new Float:_pg_ftmp;
new _human;
new _alive;
new g_max_players;
new g_msg_saytext;
new Float:g_t_origin[3];
new Float:g_b_origin[3];
new bool:g_valid_origin;
new bool:g_ml;
Float:operator-(Float:)(Float:oper)
{
	return oper ^ -2147483648;
}

Float:operator*(Float:,_:)(Float:oper1, oper2)
{
	return floatmul(oper1, float(oper2));
}

Float:operator+(Float:,_:)(Float:oper1, oper2)
{
	return floatadd(oper1, float(oper2));
}

bool:operator!=(Float:,Float:)(Float:oper1, Float:oper2)
{
	return floatcmp(oper1, oper2) != 0;
}

bool:operator<(Float:,_:)(Float:oper1, oper2)
{
	return 0 > floatcmp(oper1, float(oper2));
}

get_datadir(name[], len)
{
	return get_localinfo(0, name, len);
}

public __fatal_ham_error(Ham:id, HamError:err, reason[])
{
	new func = get_func_id(52, -1);
	new bool:fail = 1;
	new var1;
	if (func != -1 && callfunc_begin_i(func, -1) == 1)
	{
		callfunc_push_int(id);
		callfunc_push_int(err);
		callfunc_push_str(reason, 0);
		if (callfunc_end() == 1)
		{
			fail = false;
		}
	}
	if (fail)
	{
		set_fail_state(reason);
	}
	return 0;
}

public plugin_init()
{
	register_plugin(144, 236, 252);
	RegisterHam(43, 292, 340, 1);
	RegisterHam(0, 396, 424, 1);
	RegisterHam(11, 476, 504, 1);
	g_max_players = get_maxplayers();
	g_msg_saytext = get_user_msgid(560);
	new path[128];
	get_datadir(path, 127);
	add(path, 127, 592, 0);
	if (file_exists(path))
	{
		register_dictionary(724);
		g_ml = true;
	}
	return 0;
}

public button_pushed(ent, idcaller, idactivator, use_type, Float:value)
{
	new var1;
	if (!1 <= g_max_players >= idcaller || !1 << idcaller & _alive || !pev_valid(ent) || cs_get_user_team(idcaller, 0) == 1)
	{
		return 0;
	}
	new Float:ftmp[3] = 0.0;
	pev(ent, 131, 124);
	pev(ent, 132, ftmp);
	new var2 = g_b_origin;
	var2[0] = floatadd(var2[0], ftmp[0]);
	new var3 = g_b_origin[1];
	var3 = floatadd(var3, ftmp[1]);
	new var4 = g_b_origin[2];
	var4 = floatadd(var4, ftmp[2]);
	new var5 = g_b_origin;
	var5[0] = floatmul(1056964608, var5[0]);
	new var6 = g_b_origin[1];
	var6 = floatmul(1056964608, var6);
	new var7 = g_b_origin[2];
	var7 = floatmul(1056964608, var7);
	pev(idcaller, 118, 112);
	if (pev(idcaller, 84) & 16384)
	{
		g_t_origin[2] += 19;
	}
	g_valid_origin = true;
	return 0;
}

public Player_Killed(id)
{
	_alive = ~1 << id & _alive;
	return 0;
}

public Player_Spawn(id)
{
	if (!is_user_alive(id))
	{
		return 0;
	}
	_alive = 1 << id | _alive;
	if (cs_get_user_team(id, 0) != 1)
	{
		return 0;
	}
	g_valid_origin = false;
	return 0;
}

public client_putinserver(id)
{
	new var1;
	if (is_user_bot(id) || is_user_hltv(id))
	{
		return 0;
	}
	_human = 1 << id | _human;
	return 0;
}

public announce_t_change(newid, oldid)
{
	new msg[160];
	new otname[32];
	new ntname[32];
	get_user_name(oldid, otname, 31);
	get_user_name(newid, ntname, 31);
	if (!g_ml)
	{
		formatex(msg, 159, 832, otname, ntname);
		message_begin(2, g_msg_saytext, 1092, 0);
		write_byte(oldid);
		write_string(msg);
		message_end();
		return 0;
	}
	new i = 1;
	while (i <= g_max_players)
	{
		if (1 << i & _human)
		{
			formatex(msg, 159, 1104, i, 1116, otname, ntname);
			message_begin(1, g_msg_saytext, 1092, i);
			write_byte(oldid);
			write_string(msg);
			message_end();
		}
		i++;
	}
	return 0;
}

public client_disconnect(id)
{
	_human = ~1 << id & _human;
	_alive = ~1 << id & _alive;
	new players[32];
	new pnum;
	new i = 1;
	while (i <= g_max_players)
	{
		if (1 << i & _human)
		{
			if (cs_get_user_team(i, 0) == 1)
			{
				return 0;
			}
			pnum++;
			players[pnum] = i;
		}
		i++;
	}
	if (0 >= pnum)
	{
		return 0;
	}
	new new_terr = players[random(pnum)];
	cs_set_user_team(new_terr, 1, 0);
	announce_t_change(new_terr, id);
	if (!g_valid_origin)
	{
		ExecuteHamB(98, new_terr);
		return 0;
	}
	ExecuteHamB(98, new_terr);
	g_valid_origin = true;
	new param[2];
	param[0] = new_terr;
	param[1] = 10;
	restore_origin(param);
	return 0;
}

public restore_origin(param[])
{
	new var1;
	if (!g_valid_origin || var3 < 0 || !1 << param[0] & _alive)
	{
		return 0;
	}
	new tr;
	new Float:ftmp[3] = 0.0;
	ftmp[0] = g_t_origin[0];
	ftmp[1] = g_t_origin[1];
	ftmp[2] = g_t_origin[2];
	new i;
	while (i < 100)
	{
		engfunc(32, ftmp, ftmp, 0, 1, 0, tr);
		new var2;
		if (get_tr2(tr, 1) || get_tr2(tr, 0))
		{
			ftmp[0] = floatadd(ftmp[0], random_float(-1054867456, 1092616192));
			new var4 = ftmp[1];
			var4 = floatadd(var4, random_float(-1054867456, 1092616192));
			new var5 = ftmp[2];
			var5 = floatadd(var5, random_float(-1054867456, 1092616192));
			i++;
		}
		engfunc(29, 112, ftmp, 0, 0, tr);
		new Float:fraction = 0.0;
		get_tr2(tr, 4, fraction);
		if (!(1.0 != fraction))
		{
			g_t_origin[0] = ftmp[0];
			g_t_origin[1] = ftmp[1];
			g_t_origin[2] = ftmp[2];
			set_pev(param[0], 118, 112);
			ftmp[0] = -g_t_origin[0];
			ftmp[1] = -g_t_origin[1];
			ftmp[2] = -g_t_origin[2];
			ftmp[0] = floatadd(ftmp[0], g_b_origin[0]);
			new var6 = ftmp[1];
			var6 = floatadd(var6, g_b_origin[1]);
			new var7 = ftmp[2];
			var7 = floatadd(var7, g_b_origin[2]);
			_pg_ftmp = floatsqroot(floatadd(floatmul(ftmp[0], ftmp[0]), floatmul(ftmp[1], ftmp[1])));
			ftmp[1] = floatacos(floatdiv(ftmp[0], _pg_ftmp), 1) * 1 - ftmp[1] < 0 * 2;
			ftmp[0] = -floatatan(floatdiv(ftmp[2], _pg_ftmp), 1);
			ftmp[2] = 0.0;
			set_pev(param[0], 124, ftmp);
			set_pev(param[0], 65, 1);
			return 0;
		}
		set_task(1036831949, 1160, param[0] + 457547, param, 2, 1220, 0);
		return 0;
	}
	set_task(1036831949, 1160, param[0] + 457547, param, 2, 1220, 0);
	return 0;
}

